sdk/core/allocator/token_unseal.S

```diff
@@ -0,0 +1,73 @@
+#include <cheri-builtins.h>
+#include "token.h"
+
+/*
+ * void *
+ * token_obj_unseal_internal(struct SKeyStruct *ca0, struct SObjStruct *ca1)
+ *
+ * Register allocation:
+ *
+ *  - ca0 holds a sealing key, either the user's or the real deal, and is
+ *    replaced with the unsealed value or NULL
+ *
+ *  - ca1 holds the user's sealed object pointer
+ *
+ *  - t0/ct0 holds a copy of the user key
+ *
+ *  - t1/ct1 is used within each local computation and never holds secrets
+ */
+
+.global token_obj_unseal_internal
+token_obj_unseal_internal:
+
+  // Verify key tag
+  cgettag t1, ca0
+  beqz    t1, 9f
+
+  // Verify key address == base and len > 0
+  cgetbase t1, ca0 
+  bne      a0, t1, 9f // as-integer access to ca0 gives address
+  cgetlen  t1, ca0
+  beqz     t1, 9f
+
+  // Verify key has unseal permission
+  cgetperm t1, ca0
+  andi     t1, t1, CHERI_PERM_UNSEAL
+  beqz     t1, 9f
+
+  // Copy key capability to scratch register
+  cmove ct0, ca0
+
+  // Load unsealing root capability, to be clobbered by return value
+  // This faults only if something has gone very, very wrong
+1:
+  auipcc ca0, %cheri_compartment_pccrel_hi(__sealingkey)
+  clc    ca0, %cheri_compartment_pccrel_lo(1b)(ca0)
+
+  // Unseal, clobbering authority
+  cunseal ca0, ca1, ca0
+
+  // Verify tag of unsealed form
+  cgettag t1, ca0
+  beqz    t1, 9f
+
+  // Load software type.  This will not trap, thanks to above tag check.
```

Comment:
I again agitate for a conditional load instruction, predicated on the tag being set (and not updating registers if the tag is clear).

Reply:
I think we can probably add such a thing in a v2 CHERIoT ISA.  It would probably need to update the register with zero if the load failed, or it will be painful on systems with register renaming.

Reply:
Continuing the tangent:
- I don't actually want to burn instruction encoding space on such an instruction
- Jess really didn't like my HINT/NOP-based markup of instructions (the fault handler can always decode the next instruction as part of deciding what to do on the fault path), but I still think it's a decent approach.  The I$ costs are significantly lower than the costs CheriBSD on RISC-V is already paying.
- The out-of-band table thing is painful on big systems, but per-compartment handling on CHERIoT might be entirely sensible.  Though if we wanted library functions to have fault handlers (as we would, here), we'd need to pick them up from PCC, and maybe we don't want to do that.

Reply:
> I don't actually want to burn instruction encoding space on such an instruction

I think it might be pretty cheap, if we make it a masked operation.  So we just have a two-register operation that is load capability via capability, sets zeroes into any bytes of the register that you are not allowed to read.  This could then be used with narrower loads (with masking).  I don't think we need an analogous store, but if we did then it would be another two-register thing.  We could also wrap the load in a copy-from-unsafe thing that could run with interrupts enabled.